internal/sbi/consumer/nrf_service.go

```diff
@@ -0,0 +1,221 @@
+			logger.ConsumerLog.Errorf("NFInstancesStoreApi response body cannot close: %+v", resCloseErr)
+		}
+	}()
+	if res != nil && res.StatusCode == http.StatusTemporaryRedirect {
+		return nil, fmt.Errorf("Temporary Redirect For Non NRF Consumer")
+	}
+
+	return &result, nil
+}
+
+func (s *nnrfService) SendDeregisterNFInstance() (problemDetails *models.ProblemDetails, err error) {
+	logger.ConsumerLog.Infof("Send Deregister NFInstance")
+
+	ctx, pd, err := ausf_context.GetSelf().GetTokenCtx(models.ServiceName_NNRF_NFM, models.NfType_NRF)
+	if err != nil {
+		return pd, err
+	}
+
+	ausfContext := s.consumer.Context()
+	client := s.getNFManagementClient(ausfContext.NrfUri)
+
+	var res *http.Response
+
+	res, err = client.NFInstanceIDDocumentApi.DeregisterNFInstance(ctx, ausfContext.NfId)
+	if err == nil {
+		return problemDetails, err
+	} else if res != nil {
+		defer func() {
+			if resCloseErr := res.Body.Close(); resCloseErr != nil {
+				logger.ConsumerLog.Errorf("DeregisterNFInstance response cannot close: %+v", resCloseErr)
+			}
+		}()
+		if res.Status != err.Error() {
+			return problemDetails, err
+		}
+		problem := err.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
+		problemDetails = &problem
+	} else {
+		err = openapi.ReportError("server no response")
+	}
+	return problemDetails, err
+}
+
+func (s *nnrfService) RegisterNFInstance(ctx context.Context) (
+	resouceNrfUri string, retrieveNfInstanceID string, err error,
+) {
+	ausfContext := s.consumer.Context()
+
+	client := s.getNFManagementClient(ausfContext.NrfUri)
+	nfProfile, err := s.buildNfProfile(ausfContext)
+	if err != nil {
+		return "", "", errors.Wrap(err, "RegisterNFInstance buildNfProfile()")
+	}
+
+	var nf models.NfProfile
+	var res *http.Response
+	for {
+		nf, res, err = client.NFInstanceIDDocumentApi.RegisterNFInstance(context.TODO(), ausfContext.NfId, nfProfile)
+		if err != nil || res == nil {
+			logger.ConsumerLog.Errorf("AUSF register to NRF Error[%v]", err)
+			time.Sleep(2 * time.Second)
+			continue
+		}
+		defer func() {
+			if resCloseErr := res.Body.Close(); resCloseErr != nil {
+				logger.ConsumerLog.Errorf("RegisterNFInstance response body cannot close: %+v", resCloseErr)
+			}
+		}()
+		status := res.StatusCode
+		if status == http.StatusOK {
+			// NFUpdate
+			break
+		} else if status == http.StatusCreated {
+			// NFRegister
+			resourceUri := res.Header.Get("Location")
+			resouceNrfUri = resourceUri[:strings.Index(resourceUri, "/nnrf-nfm/")]
+			retrieveNfInstanceID = resourceUri[strings.LastIndex(resourceUri, "/")+1:]
+
+			oauth2 := false
+			if nf.CustomInfo != nil {
+				v, ok := nf.CustomInfo["oauth2"].(bool)
+				if ok {
+					oauth2 = v
+					logger.MainLog.Infoln("OAuth2 setting receive from NRF:", oauth2)
+				}
+			}
+			ausf_context.GetSelf().OAuth2Required = oauth2
+			if oauth2 && ausf_context.GetSelf().NrfCertPem == "" {
+				logger.CfgLog.Error("OAuth2 enable but no nrfCertPem provided in config.")
+			}
+
+			break
```

Comment:
Should we update `ausf_context` after replacing the profile in NRF?
```suggestion
		if status == http.StatusOK && status == http.StatusCreated {
			// NFRegister
			resourceUri := res.Header.Get("Location")
			resouceNrfUri = resourceUri[:strings.Index(resourceUri, "/nnrf-nfm/")]
			retrieveNfInstanceID = resourceUri[strings.LastIndex(resourceUri, "/")+1:]

			oauth2 := false
			if nf.CustomInfo != nil {
				v, ok := nf.CustomInfo["oauth2"].(bool)
				if ok {
					oauth2 = v
					logger.MainLog.Infoln("OAuth2 setting receive from NRF:", oauth2)
				}
			}
			ausf_context.GetSelf().OAuth2Required = oauth2
			if oauth2 && ausf_context.GetSelf().NrfCertPem == "" {
				logger.CfgLog.Error("OAuth2 enable but no nrfCertPem provided in config.")
			}

			break
```